more robust circleci testing

MartinNowak opened the pull request.

wilzbach commented on circle.yml:
```diff
@@ -1,21 +1,17 @@
 test:
   override:
-    - case $CIRCLE_NODE_INDEX in 0) MODEL=32 ./travis.sh ;; 1) MODEL=64 ./travis.sh ;; esac:
+    - case $CIRCLE_NODE_INDEX in 0) MODEL=32 ./circleci.sh ;; 1) MODEL=64 ./circleci.sh ;; esac:
```
@MartinNowak this is a hard-coded hack to support parallelism.
However of course, if only one container is enabled for a repo only the `32-bit` will be run.

MartinNowak replied:
It's a horrible YAML abomination b/c apparently they aren't capable to allow hashes as argument (and we don't want to use multiline bash scripts as hash key, even in YAML).

wilzbach commented on circle.yml:
```diff
@@ -1,21 +1,17 @@
   override:
-    - curl -fsSL --retry 3 "http://downloads.dlang.org/releases/2.x/${DMD_VERSION}/dmd.${DMD_VERSION}.linux.tar.xz" | tar -C ~ -Jxf -
+    - for i in {0..4}; do curl -fsS -A "$CURL_USER_AGENT" --max-time 5 https://dlang.org/install.sh -O && break || [ $i -ge 4 ] || sleep $((1 << $i)); done
+    - source $(bash install.sh dmd-$DMD_VERSION --activate)
```
Unfortunately this won't work for CircleCI has all commands are run in separate shells.

MartinNowak committed:
more robust circleci testing

- move all the logic into a separate shell script (CircleCI's YAML is
  really unfriendly for slightly more complicated stuff)
- use installer script to fetch bootstrap compiler (same retry logic as
  on Travis-CI)
- cache installed bootstrap compiler

MartinNowak committed:
rename make option to ENABLE_COVERAGE

- more consisten with other dmd make options

MartinNowak committed:
Revert "Fix CI checkout branch for non Travis CIs"

This reverts commit cffa4c3691a4323f277eca236207375302e726a0.

MartinNowak commented:
Mainly did this b/c I wanted to restructure the Travis-CI tests to start less workers, but failed due to all the extra cases for CircleCI.
Better to have that in a separate file and deal w/ only a single provider at a time.

MartinNowak committed:
remove circleci stuff from travis.sh

wilzbach commented on circleci.sh:
```diff
@@ -0,0 +1,68 @@
+case $CIRCLE_NODE_INDEX in
+    0) MODEL=64 ;;
+    1) MODEL=32 ;;
+esac
```
Could you please enable two containers, s.t. 32-bit gets tested too?

Should be here https://circleci.com/gh/dlang/dmd/edit#parallel-builds

MartinNowak replied:
We're not testing the backend for coverage, shouldn't be much difference, but well.

WalterBright replied:
I asked @wilzbach to stick with one, because the Terms Of Service only offer 4 instances for open source projects. It was unclear what CircleCI's definition of open source was, so while we await clarification I'd like to stay with one. (If their definition is "public repository", then we're golden.)

wilzbach replied:
> I asked @wilzbach to stick with one, because the Terms Of Service only offer 4 instances for open source projects. 

Did you miss this comment?

> I just found the following in the settings at the [org page](https://circleci.com/gh/organizations/dlang/settings)

>> Additionally, projects that are __public on GitHub__ will build with 3 extra containers -- our gift to free and open source software.

>and then this in their [FAQ](https://circleci.com/gh/organizations/dlang/settings#containers)

>>> What if I am building open-source?

>> We offer a total of four free linux containers ($2400 annual value) for open-source projects. __Simply keeping your project public__ will enable this for you!

https://github.com/dlang/dmd/pull/6006#issuecomment-237619185

WalterBright replied:
@wilzbach yes, I had missed that. Glad this is cleared up. Thanks!

@klickverbot Looks like we're good on that, then. But there's that other issue you mentioned - maybe throttling it by using only 1 instance will be more effective for us than running dry halfway through the month.

wilzbach replied:
> than just the latency of parallel builds on many updated PRs.

@klickverbot Let's just see whether it makes a difference and/or clogs our CircleCi queue. It's just an experiment, so we can always switch back to 1 container ;-)

wilzbach commented:
> Mainly did this b/c I wanted to restructure the Travis-CI tests to start less workers, but failed due to all the extra cases for CircleCI.
> Better to have that in a separate file and deal w/ only a single provider at a time.

Thanks a lot for this massive cleanup :+1: 
It looks a lot more professional than my hack.

MartinNowak commented on circleci.sh:
```diff
@@ -0,0 +1,68 @@
+        local base_branch=$(curl -fsSL https://api.github.com/repos/dlang/dmd/pulls/$CIRCLE_PR_NUMBER | jq -r '.base.ref')
+    else
+        local base_branch=$CIRCLE_BRANCH
+    fi
```
This is the snippet to get the target branch. Hopefully we won't hit any rate limit of github, b/c of other people on the test machine.

wilzbach replied:
I hope we don't , otherwise we can try 1) direct [authenticated access](https://developer.github.com/v3/#oauth2-keysecret) or 2) rolling our own, small API cache

MartinNowak replied:
Or extend CircleCI, but I don't it'll be a problem.

MartinNowak replied:
Let's see [How to get the pull request upstream branch? - Build Environment - CircleCI Community Discussion](https://discuss.circleci.com/t/how-to-get-the-pull-request-upstream-branch/5496).

wilzbach replied:
Btw didn't you want to merge into the current `base_branch`?
Something like simple like this should do it?

```
if [ -n ${CIRCLE_PR_NUMBER:-} ]; then
    git remote add upstream git@github.com:dlang/dmd.git
    git fetch upstream
    git merge "upstream/${base_branch}"
else
    git merge ${base_branch}
fi
```

MartinNowak replied:
Yes merging would be good.

MartinNowak commented:
Auto-merge toggled on

JackStouffer commented:
@MartinNowak This broke building dmd on OS X

```
$ make -f posix.mak AUTO_BOOTSTRAP=1
/Library/Developer/CommandLineTools/usr/bin/make -C src -f posix.mak
no cpu specified, assuming X86
mkdir -p /tmp/.host_dmd-2.068.2
curl -fsSL http://downloads.dlang.org/releases/2.x/2.068.2 /dmd.2.068.2 .osx.tar.xz | tar -C /tmp/.host_dmd-2.068.2  -Jxf - || rm -rf /tmp/.host_dmd-2.068.2
curl: (3) <url> malformed
curl: (6) Could not resolve host: .osx.tar.xz
tar: Unrecognized archive format
tar: Error exit delayed from previous errors.
CC=c++ /tmp/.host_dmd-2.068.2 /dmd2/osx/bin/dmd -conf=/tmp/ /dmd2/osx/bin/dmd.conf idgen.d
/bin/sh: /tmp/.host_dmd-2.068.2: No such file or directory
make[1]: *** [idgen] Error 127
make: *** [all] Error 2
```

JackStouffer commented on src/posix.mak:
```diff
@@ -66,7 +66,7 @@ ifeq (,$(AUTO_BOOTSTRAP))
 else
   # Auto-bootstrapping, will download dmd automatically
-  HOST_DMD_VER=2.068.2
+  HOST_DMD_VER=2.068.2 # keep in sync with other occurences of that variable, e.g. in circleci.sh
```
@MartinNowak This line is the culprit

wilzbach replied:
That's bash craziness :cry: 
@JackStouffer  do quotes work for you?

JackStouffer replied:
Deleting the space after the `2` and moving the comment to its own line works

wilzbach replied:
> Deleting the space after the 2 and moving the comment to its own line works

How about quickly submitting this as PR? ;-)
